protoc-gen-openapiv2/internal/genopenapi/naming.go

```diff
@@ -52,20 +55,46 @@ func resolveNamesLegacy(messages []string) map[string]string {
 // E.g., if the fully qualified name is `.a.b.C.D`, and there are other messages with
 // fully qualified names ending in `.D` but not in `.C.D`, it assigns the unique name `C.D`.
 func resolveNamesSimple(messages []string) map[string]string {
-	return resolveNamesUniqueWithContext(messages, 0, ".")
+	return resolveNamesUniqueWithContext(messages, 0, ".", false)
+}
+
+// resolveNamesPackage takes the names of all proto messages and generates unique references by
+// starting with the package-scoped name (with nested message types qualified by their containing
+// "parent" types), and then following the "simple" heuristic above to add package name components
+// until each message has a unique name with a "." between each component.
+//
+// E.g., if the fully qualified name is `.a.b.C.D`, the name is `C.D` unless there is another
+// package-scoped name ending in "C.D", in  which case it would be `b.C.D` (unless that also
+// conflicted, in which case the name would be the fully-qualified `a.b.C`).
+func resolveNamesPackage(messages []string) map[string]string {
+	return resolveNamesUniqueWithContext(messages, 0, ".", true)
 }
 
 // Take the names of every proto message and generates a unique reference by:
 // first, separating each message name into its components by splitting at dots. Then,
 // take the shortest suffix slice from each components slice that is unique among all
 // messages, and convert it into a component name by taking extraContext additional
 // components into consideration and joining all components with componentSeparator.
-func resolveNamesUniqueWithContext(messages []string, extraContext int, componentSeparator string) map[string]string {
+func resolveNamesUniqueWithContext(messages []string, extraContext int, componentSeparator string, qualifyNestedMessages bool) map[string]string {
 	packagesByDepth := make(map[int][][]string)
 	uniqueNames := make(map[string]string)
 
 	hierarchy := func(pkg string) []string {
-		return strings.Split(pkg, ".")
+		if !qualifyNestedMessages {
+			return strings.Split(pkg, ".")
+		}
+		// We rely on the convention that package names are lowercase but message names are
+		// capitalized.
+		pkgEnd := regexp.MustCompile(`\.[A-Z]`).FindStringIndex(pkg)
```

Comment:
Could we move this regexp compilation into a global variable?

Reply:
Sure!

I actually had it as a global var initially, as is often done for regexp compilation... but I moved it into the function to have the pattern defined in the same context as it it used for readability. Generally name resolution only happens once per process anyways, so I don't think the performance optimization of doing the regexp compilation one time up-front really gains us anything here.

I've moved it out though into a global var though, let me know what you think.

Reply:
Thanks, I appreciate the explanation. All the same, I think anyone reading an inline `regexp.MustCompile` will be surprised, so lets leave it as a package level variable.